OpenRA.Mods.Common/Widgets/Logic/Lobby/MapPreviewLogic.cs

```diff
@@ -44,220 +51,258 @@ public class MapPreviewLogic : ChromeLogic
+				var truncateCache = new CachedTransform<MapPreview, string>(
+					m => WidgetUtils.TruncateText(authorCache.Update(m.Author), authorLabel.Bounds.Width, font));
+
+				authorLabel.GetText = () => truncateCache.Update(getMap().Map);
+
+				return parent;
 			}
 
-			var progress = widget.GetOrNull("MAP_PROGRESS");
-			if (progress != null)
+			void SetUpInstallButton(Widget parent)
 			{
-				progress.IsVisible = () =>
+				var button = parent.Get<ButtonWidget>("MAP_INSTALL");
+				button.IsHighlighted = () => blink;
+				button.OnClick = () =>
 				{
-					var (map, _) = getMap();
-					return map.Status != MapStatus.Available && map.Status != MapStatus.DownloadAvailable;
+					getMap().Map.Install(mapRepository, () =>
+					{
+						if (orderManager != null)
+							Game.RunAfterTick(() => orderManager.IssueOrder(Order.Command($"state {Session.ClientState.NotReady}")));
+					});
 				};
+			}
 
-				SetupWidgets(progress, modData, getMap, onMouseDown, getSpawnOccupants, getDisabledSpawnPoints, showUnoccupiedSpawnpoints);
+			previewWidgets[PreviewStatus.Unknown] = new Widget[] { previewLarge };
+			previewWidgets[PreviewStatus.Playable] = new Widget[] { previewLarge, SetupAuthorAndMapType("MAP_AVAILABLE") };
+			previewWidgets[PreviewStatus.Incompatible] = new Widget[] { previewLarge, widget.Get("MAP_INVALID") };
+			previewWidgets[PreviewStatus.Validating] = new Widget[] { previewSmall, widget.Get("MAP_VALIDATING") };
 
-				var statusSearching = progress.GetOrNull("MAP_STATUS_SEARCHING");
-				if (statusSearching != null)
-				{
-					statusSearching.IsVisible = () =>
-					{
-						var (map, _) = getMap();
-						return map.Status == MapStatus.Searching;
-					};
-				}
+			var donwloadableContainer = SetupAuthorAndMapType("MAP_DOWNLOADABLE");
+			SetUpInstallButton(donwloadableContainer);
+			previewWidgets[PreviewStatus.DownloadAvailable] = new Widget[] { previewSmall, donwloadableContainer };
 
-				var statusUnavailable = progress.GetOrNull("MAP_STATUS_UNAVAILABLE");
-				if (statusUnavailable != null)
-				{
-					statusUnavailable.IsVisible = () =>
-					{
-						var (map, _) = getMap();
-						return map.Status == MapStatus.Unavailable && map != MapCache.UnknownMap;
-					};
-				}
+			var updateButton = widget.Get<ButtonWidget>("MAP_UPDATE");
+			updateButton.IsHighlighted = () => blink;
+			updateButton.OnClick = () =>
+			{
+				var uid = modData.MapCache.GetUpdatedMap(getMap().Map.Uid);
+				if (uid != null)
+					onMapUpdate(uid);
+			};
 
-				var statusError = progress.GetOrNull("MAP_STATUS_ERROR");
-				if (statusError != null)
-					statusError.IsVisible = () => getMap().Map.Status == MapStatus.DownloadError;
+			var updatableDownloadableContainer = widget.Get("MAP_DOWNLOADABLE_UPDATABLE");
+			SetUpInstallButton(updatableDownloadableContainer);
+			previewWidgets[PreviewStatus.UpdateDownloadAvailable] = new Widget[] { previewSmall, updatableDownloadableContainer, updateButton };
```

Comment:
uid doesn't change, MapPreview.Status changes

Reply:
It's a bit confusing because statuses are everywhere. There were `MapPreview.Status`, Session.MapStatus` and now `MapPreviewLogic.PreviewStatus` was added

Reply:
Is it really necessary to introduce this new enum? You yourself say that it's confusing to have different status variables everywhere. Previously you could immediately see when a widget is visible and when it won't be. Now you have to 1) know that you have to look inside `Tick`, 2) follow a switch case and then 3) an if else branching to cover all the logic, to then 4) see which widgets are set to visible in `UpdateVisibility`. I don't see the benefit to be completely honest.

Reply:
> Previously you could immediately see when a widget is visible and when it won't be
 
 But you couldn't see what widgets were visible / active and when. With my system all you need to know is in what previewWidgets dictionary the widget is

Reply:
And then you need to know which previewWidgets dictionary is active... This boils down to the same as previously just with extra steps/overhead.

Reply:
to know which one is active you only need to know `PreviewStatus currentStatus`

Reply:
Yes, and that means you need to go through the four steps I outlined above if you want to know when for example the download widget will be active.

Reply:
True, it's still complex, but switch cases in a single place are much easier to follow than a bunch of complex if statements scattered everywhere among the code. One of the main things this implementation allows us is it to trigger the Retry button on relevant status changes, something that would require a bunch of duplication and confusing code with the old way.

Reply:
Ok, that sounds reasonable.